dmriprep/utils/vectors.py

```diff
@@ -1,482 +1,346 @@
+    ...          (0.0, 0.0, 1.0)]  # Just half a sphere
+    >>> calculate_pole(bvecs).tolist()
+    [0.0, 0.0, 1.0]
+
+    >>> bvecs += [(0.0, 0.0, -1.0)]  # Make it a full-sphere
+    >>> calculate_pole(bvecs).tolist()
+    [0.0, 0.0, 0.0]
 
     References
     ----------
     https://rstudio-pubs-static.s3.amazonaws.com/27121_a22e51b47c544980bad594d5e0bb2d04.html
+
     """
-    import itertools
-    if vecs.shape[1] != 3:
-        raise ValueError("Input vectors must be 3D vectors")
-    if not np.allclose(1, np.linalg.norm(vecs, axis=1)):
-        raise ValueError("Input vectors must be unit vectors")
+    bvecs = np.array(bvecs, dtype='float32')  # Normalize inputs
+    b0s = np.linalg.norm(bvecs, axis=1) < bvec_norm_epsilon
 
+    bvecs = bvecs[~b0s]
     # Generate all pairwise cross products.
-    v0, v1 = zip(*[p for p in itertools.permutations(vecs, 2)])
-    cross_prods = np.cross(v0, v1)
+    pairs = np.array(list(permutations(bvecs, 2)))
+    pairs = np.swapaxes(pairs, 0, 1)
+    cross_prods = np.cross(pairs[0, ...], pairs[1, ...])
 
     # Normalize them.
-    cross_prods /= np.linalg.norm(cross_prods, axis=1)[:, np.newaxis]
-
-    # `cross_prods` now contains all candidate vertex points for "the polygon" in the reference. "The polygon" is a
-    # subset. Find which points belong to the polygon using a dot product test with each of the original vectors.
-    angles = np.arccos(np.dot(cross_prods, vecs.transpose()))
+    cross_norms = np.linalg.norm(cross_prods, axis=1)
+    cross_zeros = cross_norms < 1.0e-4
+    cross_prods = cross_prods[~cross_zeros]
+    cross_prods /= cross_norms[~cross_zeros, np.newaxis]
+
+    # `cross_prods` now contains all candidate vertex points for "the polygon"
+    # in the reference.
+    # "The polygon" is a subset.
+    # Find which points belong to the polygon using a dot product test
+    # with each of the original vectors.
+    angles = np.arccos(cross_prods.dot(bvecs.T))
 
     # And test whether it is orthogonal or less.
     dot_prod_test = angles <= np.pi / 2.0
+    ntests = dot_prod_test.sum(axis=1) == len(bvecs)
 
-    # If there is at least one point that is orthogonal or less to each input vector, then the points lie on some
-    # hemisphere.
-    is_hemi = len(vecs) in np.sum(dot_prod_test.astype(int), axis=1)
+    # If there is at least one point that is orthogonal or less to each
+    # input vector, then the points lie on some hemisphere.
+    is_hemi = np.any(ntests)
 
+    pole = np.zeros(3)
     if is_hemi:
-        vertices = cross_prods[np.sum(dot_prod_test.astype(int), axis=1) == len(vecs)]
+        vertices = cross_prods[ntests]
         pole = np.mean(vertices, axis=0)
         pole /= np.linalg.norm(pole)
-    else:
-        pole = np.array([0.0, 0.0, 0.0])
-    return is_hemi, pole
-
-
-def set_slm(gtab, is_hemi):
-    """
-    Evaluate eddy compatibility and set second level model that specifies the mathematical form for how the
```

Comment:
No need for this. We can add a function node right before eddy to check this.

Reply:
Now that the `pole` output parameter is available on the interface, doing this will be easy. It was drafted as an 'all-in-one' function before since there was not way otherwise to keep the hemisphericity information available to downstream eddy-related nodes in the workflow. I suppose I'm seeing some of the value in nipype interfaces now!